roboteam_ai/src/utilities/OffensiveCoach.cpp

```diff
@@ -0,0 +1,312 @@
+//
+// Created by robzelluf on 3/8/19.
+//
+
+#include <roboteam_ai/src/interface/widget.h>
+#include <roboteam_ai/src/interface/drawer.h>
+#include "OffensiveCoach.h"
+
+namespace rtt {
+namespace ai {
+namespace coach {
+
+int OffensiveCoach::maxPositions = 4;
+double OffensiveCoach::maxDistanceFromBall = 6.0;
+double OffensiveCoach::marginFromLines = 0.2;
+std::vector<OffensiveCoach::OffensivePosition> OffensiveCoach::offensivePositions;
+std::map<int, OffensiveCoach::OffensivePosition> OffensiveCoach::robotPositions;
+
+double OffensiveCoach::calculateCloseToGoalScore(Vector2 position) {
+    double distanceFromGoal = (Field::get_their_goal_center() - position).length();
+
+    double score = exp(-0.1 * distanceFromGoal);
+    return score;
+}
+
+double OffensiveCoach::calculateShotAtGoalScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
+    while (safelyness > 0) {
+            if (control::ControlUtils::clearLine(position, Field::get_their_goal_center(), world, safelyness)) {
+                break;
+            }
+            safelyness -= 0.5;
+        }
+
+    return 1 - exp(-safelyness);
+}
+
+
+double OffensiveCoach::calculatePassLineScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
+    while (safelyness > 0) {
+        if (control::ControlUtils::clearLine(world.ball.pos, position, world, safelyness)) {
+            break;
+        }
+        safelyness -= 0.5;
+    }
+
+    return 1 - exp(-safelyness);
+}
+
+double OffensiveCoach::calculateDistanceToOpponentsScore(Vector2 position, roboteam_msgs::World world) {
+    shared_ptr<roboteam_msgs::WorldRobot> closestRobot = World::getRobotClosestToPoint(world.them, position);
+    if (closestRobot) {
+        double distance = (position - closestRobot->pos).length();
```

Comment:
the whole safelyness is magic

Reply:
I agree, but I also think it's an okay method. It's just a measure for the degree the line is not obstructed by other objects